revise Stratum 1 part (#9)
cvmfs-contrib/cvmfs-tutorial-2021: docs/03_stratum1_proxies.md

Therefore, we will show how all these points can be addressed by adding a **Stratum 1 server** and a **caching proxy server**.

Quick reminder: a Stratum 1 is a *replica server* that keeps *mirrors* of the repositories served by a Stratum 0.
It is a web server that periodically synchronizes the contents of the repositories.

In contrast to the central Stratum 0 server you can have *multiple* Stratum 1 servers,
and it is recommended to have them geographically distributed, so that clients always have a nearby Stratum 1.

How many Stratum 1 servers you need mostly depends on the number of clients and how they are
distributed geographically, but often a few is already sufficient.

Scalability and performance can be improved with proxies, which we will discuss later in this section.

<p align="center">
<img src="../img/stratum1_proxy.png" alt="CernVM-FS - Stratum 1 + proxy" width="700px"/>
</p>

## 3.1 Setting up the Stratum 1 server

### 3.1.1 Requirements

A Stratum 1 server has [similar requirements as a Stratum 0 in terms of resources](02_stratum0_client.md#211-requirements).

> In addition to port 80 (for the Apache web server0, port 8000 also has to be accessible for a Stratum 1

Reviewer: ```suggestion
In addition to port 80 (for the Apache web server), port 8000 also has to be accessible for a Stratum 1
```

cache_mem 128 MB
```

To clarify:

- [`http_port`](http://www.squid-cache.org/Doc/config/http_port) specifies on which ports Squid will listen for HTTP requests;

> - [`http_access`](http://www.squid-cache.org/Doc/config/http_access) specifies access restrictions for HTT traffic (none, in this case);

Reviewer: ```suggestion
- [`http_access`](http://www.squid-cache.org/Doc/config/http_access) specifies access restrictions for HTTP traffic (none, in this case);
```

  paths that match the regular expression on the line above);
- [`cache_mem`](http://www.squid-cache_mem.org/Doc/config/cache_mem) specifies the amount of memory that Squid is allowed to use;

For more information, see the Squid documentation: [http://www.squid-cache.org/Doc/config/](http://www.squid-cache.org/Doc/config/).

#### Start & enable services

Finally, we start and enable both the Apache and Squid services:

```bash
sudo systemctl start httpd
sudo systemctl start squid
sudo systemctl enable httpd
sudo systemctl enable squid
```

### 3.1.4 DNS cache

As a Stratum 1 server does a lot of DNS lookups, it is recommended to have a local DNS caching server on that same
system.

We will not discuss this topic any further here, but you can use `dnsmasq`, `bind`, or `systemd-resolved`.
See for instance [this tutorial](https://geekflare.com/linux-server-local-dns-caching/) for setting up `systemd-resolved`.

### 3.1.5 Creating the Stratum 1 replica

With all the required components in place, we can now really set up our Stratum 1 replica server.

#### Create and add Geo API key (optional)

> We first add our Geo API key to the CernVM-FS server settings, by creating it via

Reviewer: ```suggestion
We first add our Geo API key to the CernVM-FS server settings, by creating it
```
?

`/etc/cvmfs/keys/organization.tld/` (note the extra level!) on the Stratum 1 server, just like we did on the client.

#### Create replica

Now we make the replica by giving the URL to the repository on the Stratum 0 server
(which is always like `http://host:port/cvmfs/repository`, with the `:port` part optional)
and the path to the corresponding public master key:

```bash
sudo cvmfs_server add-replica -o $USER http://<STRATUM0_IP>/cvmfs/repo.organization.tld /etc/cvmfs/keys/organization.tld/
```

***Replace the ``<STRATUM0_IP>`` part with the IP address of your Stratum 0 server,
and adjust for the name and domain of your CernVM-FS repository!***

Executing the `add-replica` command should produce output reporting on the steps being performed,
and only take a couple of moments to complete.

*If no output is produced and the command seems to be hanging, make sure that port 80 on your Stratum 0 server
is accessible via the IP address you are using!*

#### Bypassing the Geo API license key

If you prefer not to create a MaxMind account and Geo API license key for the sake of this tutorial,
you can bypass the "`CVMFS_GEO_LICENSE_KEY not set`" error message

> produced by `cvmfs_server add-replica` by defining the `$CVMFS_GEO_DB_FILE` environment variable
> to `NONE` before running the command:
> 
> ```bash
> # only do this if you do not want to provide a Geo API key (not recommended!)
> export CVMFS_GEO_DB_FILE=NONE

Reviewer: ```suggestion
produced by `cvmfs_server add-replica` by setting the server variable `CVMFS_GEO_DB_FILE` 
to `NONE` before running the command:

```bash
# only do this if you do not want to provide a Geo API key (not recommended!)
echo 'CVMFS_GEO_DB_FILE=NONE' | sudo tee -a /etc/cvmfs/server.local
```

sudo systemctl enable squid
```

## 3.3 Re-configuring the client

Now that we have a Stratum 0 server, a Stratum 1 server, and a Squid proxy,
we have the (minimal) infrastructure in place for a production-ready CernVM-FS setup.

So we can now configure our client properly, and start using the repository.

In the previous section we connected our client directly to the Stratum 0. We are going to reuse that configuration, and only need to change two things.

### 3.3.1 Connect to the Stratum 1

We used the URL of the Stratum 0 in the file `/etc/cvmfs/config.d/repo.organization.tld.conf`. We should now change this URL, and point to the Stratum 1 instead:
```
CVMFS_SERVER_URL="http://<STRATUM1_IP>/cvmfs/@fqrn@"
```

***Replace the ``<STRATUM1_IP>`` part with the IP address of your Stratum 1 server!***

When you have more Stratum 1 servers inside the organization, you can make it a semicolon-separated list of servers. The Geo API will make sure that your client always connects to the geographically closest Stratum 1 server.

### 3.3.2 Use the Squid proxy

In order to use the local cache layer of our proxy, we have to instruct the client to send all requests through the proxy.

> This needs two small change in `/etc/cvmfs/default.local`, where you will have to replace `DIRECT`

Reviewer: ```suggestion
This needs one small change in `/etc/cvmfs/default.local`, where you will have to replace `DIRECT`
```
or (if you count the port number as second change):
```suggestion
This needs two small changes in `/etc/cvmfs/default.local`, where you will have to replace `DIRECT`
```
